opencensus/common/schedule/__init__.py

```diff
@@ -50,3 +52,121 @@ def run(self):
 
     def cancel(self):
         self.finished.set()
+
+
+class QueueEvent(object):
+    def __init__(self, name):
+        self.name = name
+        self.event = threading.Event()
+
+    def __repr__(self):
+        return ('{}({})'.format(type(self).__name__, self.name))
+
+    def set(self):
+        return self.event.set()
+
+    def wait(self, timeout=None):
+        return self.event.wait(timeout)
+
+
+class QueueExitEvent(QueueEvent):
+    pass
+
+
+class Queue(queue.Queue):
+    def __init__(self, capacity):
+        self.EXIT_EVENT = QueueExitEvent('EXIT')
+        self._queue = queue.Queue(maxsize=capacity)
+
+    def _gets(self, count, timeout):
+        start_time = time.time()
+        elapsed_time = 0
+        cnt = 0
+        while cnt < count:
+            try:
+                item = self._queue.get(block=False)
+                yield item
+                if isinstance(item, QueueEvent):
+                    return
+            except queue.Empty:
+                break
+            cnt += 1
+        while cnt < count:
+            wait_time = max(timeout - elapsed_time, 0)
+            try:
+                item = self._queue.get(block=True, timeout=wait_time)
+                yield item
+                if isinstance(item, QueueEvent):
+                    return
+            except queue.Empty:
+                break
+            cnt += 1
+            elapsed_time = time.time() - start_time
+
+    def gets(self, count, timeout):
+        return tuple(self._gets(count, timeout))
+
+    def flush(self, timeout=None):
+        start_time = time.time()
+        wait_time = timeout
+        event = QueueEvent('SYNC(timeout={})'.format(wait_time))
+        try:
+            self._queue.put(event, block=True, timeout=wait_time)
+        except queue.Full:
+            return
+        elapsed_time = time.time() - start_time
+        wait_time = timeout and max(timeout - elapsed_time, 0)
+        if event.wait(timeout):
```

Comment:
I think only the exporter would know when the batch got handled, instead of relying on the return of `export`? This gives the flexibility for exporter to do async work.

Reply:
I see, but the only reason you need the event is to get the duration? It seems better to me not to force exporters to handle the event, especially if the event would be internal to the queue otherwise.

What do you imagine using the duration for?

Reply:
> I see, but the only reason you need the event is to get the duration?

Besides duration, the exporter could tell if there is an explicit intention to flush the queue, or it is under situation where the application is exiting.

> It seems better to me not to force exporters to handle the event, especially if the event would be internal to the queue otherwise.

Yep, it is better not to force every exporter to handle the event, unless they explicitly ask for it (and benefit from it).

> What do you imagine using the duration for?

I think the duration is less important than the event itself. The event could be useful for telling whether we're about to exit, or there is explicit intention to flush.

Reply:
If the goal is to let exporters do some cleanup work on shutdown, what do you think about having a separate `shutdown` API method instead of the event? 

@bogdandrutu suggests an interface like this, implemented by the worker + queue class and the exporters:

``` 
interface SpanConsumer

void addSpans(List<Span>)
# add all spans to the queue

void shutdown()
# 1. stop accepting new spans (addSpans now raises)
# 2. flush existing spans
# 3. call shutdown on next consumer in pipeline
# 4. cleanup work, e.g. shut down worker thread
```

For the exporters `addSpans` is `export`.

Reply:
Trying to understand the proposal here:
1. Looks like `addSpans` will run under the users' context, and return immediately after it enqueues the spans (or discard the spans if queue is full).
2. `shutdown` is a blocking API, which does the cleanup work, and return after done the cleanup. (will shutdown has an input called remaining_time, or it could block indefinitely?)
3. How do we plan to do flush?

Reply:
For (2) passing a timeout down the chain seems like it could work. I see how a long blocking call could be a problem here.

For (3) shutdown is effectively a flush -- the queue would send all spans to the exporter, the exporter would try to send all spans to the backend. Where would you expect to call flush?

Reply:
One benefit of `flush` is that users would know if it returns true, the telemetry data is safe. The application can also decide if it should proceed or stop if the telemetry is important (e.g. auditing events).